rectools/models/candidate_ranking.py

```diff
@@ -0,0 +1,637 @@
+# flake8: noqa
+# TODO: docstrings
+
+import typing as tp
+from functools import reduce
+
+import attr
+import numpy as np
+import pandas as pd
+from catboost import CatBoostClassifier, CatBoostRanker, Pool
+
+from rectools import Columns
+from rectools.dataset import Dataset
+from rectools.dataset.identifiers import ExternalIds
+from rectools.model_selection import Splitter
+from rectools.models.base import ErrorBehaviour, ModelBase, NotFittedError
+
+
+@tp.runtime_checkable
+class ClassifierBase(tp.Protocol):
+    def fit(self, *args: tp.Any, **kwargs: tp.Any) -> None: ...
+
+    def predict_proba(self, *args: tp.Any, **kwargs: tp.Any) -> np.ndarray: ...
+
+
+@tp.runtime_checkable
+class RankerBase(tp.Protocol):
+    def fit(self, *args: tp.Any, **kwargs: tp.Any) -> None: ...
+
+    def predict(self, *args: tp.Any, **kwargs: tp.Any) -> np.ndarray: ...
+
+
+class Reranker:
+    def __init__(
+        self,
+        model: tp.Union[ClassifierBase, RankerBase],
+        fit_kwargs: tp.Optional[tp.Dict[str, tp.Any]] = None,
+    ):
+        self.model = model
+        self.fit_kwargs = fit_kwargs
+
+    def prepare_fit_kwargs(self, candidates_with_target: pd.DataFrame) -> tp.Dict[str, tp.Any]:
+        candidates_with_target = candidates_with_target.drop(columns=Columns.UserItem)
+
+        fit_kwargs = {
+            "X": candidates_with_target.drop(columns=Columns.Target),
+            "y": candidates_with_target[Columns.Target],
+        }
+
+        if self.fit_kwargs is not None:
+            fit_kwargs.update(self.fit_kwargs)
+
+        return fit_kwargs
+
+    def fit(self, candidates_with_target: pd.DataFrame) -> None:
+        fit_kwargs = self.prepare_fit_kwargs(candidates_with_target)
+        self.model.fit(**fit_kwargs)
+
+    def rerank(self, candidates: pd.DataFrame) -> pd.DataFrame:
+        reco = candidates[Columns.UserItem].copy()
+        x_full = candidates.drop(columns=Columns.UserItem)
+
+        # TODO: think about it
+        # x_full = x_full[self.model.feature_names_]
+
+        if isinstance(self.model, ClassifierBase):
+            reco[Columns.Score] = self.model.predict_proba(x_full)[:, 1]
+        elif isinstance(self.model, RankerBase):
+            reco[Columns.Score] = self.model.predict(x_full)
+        else:
+            raise ValueError("Got unexpected model_type")
+        reco.sort_values(by=[Columns.User, Columns.Score], ascending=False, inplace=True)
```

Comment:
I think it'd be good to keep the order of users. At least we keep it in other models afaik

Reply:
Fixed

Reply:
Can't see a way to make it faster. This implementation is still the fastest out of alternatives.

Reply:
First, it worth to note that in your current solution user order is not preserving, you need to add `sort=False` in the `groupby` method. Also please add a test on this.

Second, I'm getting warning (pandas version is `2.2.0`)
<img width="1019" alt="image" src="https://github.com/user-attachments/assets/4a77f97e-b6c2-450f-b160-eed48d3ab495">

Reply:
I tried couple of solutions, they may be faster with medium number of candidates per user (<100) but are slow with bigger numbers.

However I've checked polars and it's really good here. I think it may make sense to add it since sorting in pandas is really slow. wdyt?
<img width="811" alt="image" src="https://github.com/user-attachments/assets/e1c73c60-e5af-48bb-a2c8-a7cd86e5d2c4">

Reply:
And one more important thing: we very likely can speed this up independently of chosen method using `partition` algorithm. Instead of sorting 300 candidates and selecting then top-10, it's much faster to pick 10 largest scores and then sort them, as we did before the 'ImplicitRanker` for vector models.